ale_linters/go/gobuild.vim

```diff
@@ -1,214 +1,51 @@
-    endif
+    " Get a listing of all go files in the directory.
+    " TODO: Handle packages that contain c files.
+    let l:all_files = globpath(l:this_package, '*.go', 1, 1)
 
-    " only consider buffers other than a:buffer if they have the same import
-    " path as a:buffer and are modified
-    if l:bufnum != a:buffer
-      if s:PackageImportPath(l:bufnum) !=# l:importpath
-        continue
-      endif
-
-      if !getbufvar(l:bufnum, '&mod')
-        continue
-      endif
-    endif
-
-    call writefile(getbufline(l:bufnum, 1, '$'), l:tempdir . '/src/' . s:PkgFile(l:bufnum))
-  endfor
-
-  let l:gopaths = [ l:tempdir ]
-  call extend(l:gopaths, split(s:go_env.GOPATH, s:SplitChar))
-
-  return 'GOPATH=' . shellescape(join(l:gopaths, s:SplitChar)) . ' go test -c -o /dev/null ' . shellescape(l:importpath)
-endfunction
-
-function! s:PkgFile(buffer) abort
-  let l:bufname = resolve(bufname(a:buffer))
-  let l:importpath = s:PackageImportPath(a:buffer)
-  let l:fname = fnamemodify(l:bufname, ':t')
-
-  return l:importpath . '/' . l:fname
-endfunction
-
-function! s:FindBuffer(file) abort
-  for l:buffer in range(1, bufnr('$'))
-    if !buflisted(l:buffer)
-      continue
-    endif
-
-    let l:pkgfile = s:PkgFile(l:buffer)
-
-    if a:file =~ '/' . l:pkgfile . '$'
-      return l:buffer
-    endif
-  endfor
-
-  return -1
-endfunction
-
-let s:path_pattern = '[a-zA-Z]\?\\\?:\?[[:alnum:]/\.\-_]\+'
-let s:handler_pattern = '^\(' . s:path_pattern . '\):\(\d\+\):\?\(\d\+\)\?: \(.\+\)$'
-
-let s:multibuffer = 0
-
-function! ale_linters#go#gobuild#Handler(buffer, lines) abort
-  let l:output = []
-
-  for l:line in a:lines
-    let l:match = matchlist(l:line, s:handler_pattern)
-
-    if len(l:match) == 0
-      continue
-    endif
-
-    let l:buffer = s:FindBuffer(l:match[1])
-
-    if l:buffer == -1
-      continue
-    endif
+    " Filter out the current file since we don't want to include it twice.
+    " We'll then pass this list to go compile and stdin_wrapper will add on
+    " the temporary version of the current file.
+    let l:this_file = s:ThisFile(a:buffer)
+    call filter(l:all_files, 'fnamemodify(v:val, '':t'') != l:this_file')
+    call filter(l:all_files, 'fnamemodify(v:val, '':t'') !~ ''ale\.[0-9]\+\..*\.go''')
 
-    if !s:multibuffer && l:buffer != a:buffer
-      " strip lines from other buffers
-      continue
+    " Write current buffer to a temporary file.
+    let l:temp_file = s:TempFileName(a:buffer)
+    if filereadable(l:temp_file)
+      " File already exists.  Bail.
+      return ''
     endif
 
-    call add(l:output, {
-    \   'bufnr': l:buffer,
-    \   'lnum': l:match[2] + 0,
-    \   'vcol': 0,
-    \   'col': l:match[3] + 0,
-    \   'text': l:match[4],
-    \   'type': 'E',
-    \   'nr': -1,
-    \})
-  endfor
+    call writefile(getbufline(a:buffer, 1, '$'), l:temp_file)
 
-  return l:output
+    return 'go test -c ' . ' -o /dev/null ' . join(l:all_files) . ' ' . l:temp_file . '; rm ' . l:temp_file
```

Comment:
Yep,  totally aware of these restrictions.  As mentioned this PR is a very hastily put together demonstration of the concept.
Seems like it might have legs,  so I'll go ahead and flesh it out a bit. 😄